File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/handlers/CreateGameHandler.java
package systems.zlink.samples.tictactoe.server.play.handlers;

import java.util.concurrent.CompletionStage;
import org.springframework.beans.factory.ObjectProvider;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.framework.handlers.ZLinkRequest;
import systems.zlink.framework.spots.ZLinkSpotManager;
import systems.zlink.samples.tictactoe.server.configuration.SampleNames;
import systems.zlink.samples.tictactoe.server.configuration.SampleSettings;
import systems.zlink.samples.tictactoe.server.play.gamespots.TicTacToeGame;
import systems.zlink.samples.tictactoe.shared.contracts.CreateGameReq;
import systems.zlink.samples.tictactoe.shared.contracts.CreateGameRes;

@ZLinkHandlerGroup(SampleNames.PlayChannel)
public final class CreateGameHandler {
    private final ObjectProvider<ZLinkSpotManager> spots;
    private final SampleSettings settings;

    public CreateGameHandler(ObjectProvider<ZLinkSpotManager> spots, SampleSettings settings) {
        this.spots = spots;
        this.settings = settings;
    }

    @ZLinkRequest(packetName = "CreateGameReq")
    public CompletionStage<CreateGameRes> createAsync(CreateGameReq request) {
        return spots.getObject().createAsync(TicTacToeGame.class)
            .thenApply(created -> new CreateGameRes(
                created.spotRid().toHex(),
                settings.playEndpoint(),
                request.gameName()));
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/handlers/CreateGameHandler.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.handlers

import kotlinx.coroutines.future.await
import org.springframework.beans.factory.ObjectProvider
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.handlers.ZLinkRequest
import systems.zlink.framework.spots.ZLinkSpotManager
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleNames
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleSettings
import systems.zlink.samples.kotlin.tictactoe.server.play.gamespots.TicTacToeGame
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.CreateGameReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.CreateGameRes

@ZLinkHandlerGroup(SampleNames.PlayChannel)
class CreateGameHandler(
    private val spots: ObjectProvider<ZLinkSpotManager>,
    private val settings: SampleSettings,
) {
    @ZLinkRequest(packetName = "CreateGameReq")
    suspend fun create(request: CreateGameReq): CreateGameRes {
        val created = spots.getObject().createAsync(TicTacToeGame::class.java).await()
        return CreateGameRes(
            gameId = created.spotRid().toHex(),
            playEndpoint = settings.playEndpoint,
            gameName = request.gameName,
        )
    }
}
